src/ExifViewer.jl

```diff
@@ -32,75 +56,80 @@ Reads EXIF data from a filepath, IO, vector{Int8} and returns a Dict with tags a
-  ⋮                            => ⋮
-
-```
-
-
-```julia
-julia> using BenchmarkTools, TestImages, ExifViewer
-julia> filename = "test/test_images/test.jpg"
-
-julia> io = open(filename, "r")
-julia> @btime read_metadata(io)
-1.862 μs (11 allocations: 912 bytes)
-
-julia> @btime read_metadata(filename)
-269.472 μs (122 allocations: 100.05 KiB)
-
-julia> @btime data = read(filename)
-15.876 μs (13 allocations: 69.28 KiB)
-
-julia> @btime read_metadata(data)
-221.786 μs (109 allocations: 30.77 KiB)
+      "EXIF_TAG_FLASH_PIX_VERSION" => "FlashPix Version 1.0"
+      "EXIF_TAG_ORIENTATION"       => "Top-left"
+      "EXIF_TAG_EXIF_VERSION"      => "Exif Version 2.1\0ion"
+      ⋮                            => ⋮
+    
 ```
 """
-function read_metadata(data::Vector{UInt8}; ifds::Union{Int, NTuple, UnitRange}=1)
+function read_metadata(
+    data::Vector{UInt8};
+    allifds = true,
+    ifds::Union{Int,NTuple,UnitRange} = 1,
+    extract_thumbnail = false,
+)
     ed_ptr = LibExif.exif_data_new_from_data(data, length(data))
-    # handle case where ed_ptr is null
-    if(ed_ptr == C_NULL)
-        return error("Unable to read EXIF data")
+    if (ed_ptr == C_NULL)
+        return error("Unable to read EXIF data: invalid pointer")
     end
     ed = unsafe_load(ed_ptr)
-    ifds = collect(1:numifds(ed)) # reads all
-    # ifds = collect(ifds) # assumes user knows what they are doing
-    result = Dict()
+
+    if (allifds == true)
+        ifds = collect(1:numifds(ed))
+    else
+        ifds = collect(ifds)
+    end
+
+    result = Dict{String,Any}()
+    str = Vector{Cuchar}(undef, 1024)
     for i in ifds
         content_ptr = ed.ifd[i] # ques: do we need to unref these too?
         # handle case where content_ptr is null
-        if(content_ptr == C_NULL)
-            return error("Unable to read IFD:",i)
+        if (content_ptr == C_NULL)
+            return error("Unable to read IFD:", i)
         end
         data = unsafe_load(content_ptr)
+        if data.count == 0
+            continue
+        end
         res = unsafe_wrap(Array, data.entries, data.count)
-        for i in 1:data.count
-            str = Vector{Cuchar}(undef, 1024);
-            LibExif.exif_entry_get_value(Ref(unsafe_load(res[i])), str, length(str))
-            result[string(unsafe_load(res[i]).tag)] = rstrip(String(str), '\0')
+        for i = 1:data.count
+            entry = unsafe_load(res[i])
+            LibExif.exif_entry_get_value(Ref(entry), str, length(str))
+            tag = String(copy(str))[1:findfirst(iszero, str)-1]
+            tag = fixformat(tag, entry.format)
+            result[string(entry.tag)] = tag
         end
     end
-    LibExif.exif_data_unref(ed_ptr);
+
+    if (extract_thumbnail == true)
+        thumbnail_size = Int(ed.size)
+        thumbnail_data = unsafe_wrap(Array, ed.data, thumbnail_size)
+        result["EXIF_TAG_THUMBNAIL_DATA"] = thumbnail_data
+    end
+
+    LibExif.exif_data_unref(ed_ptr)
```

Comment:
For safety, this should be wrapped as "try... finally" block. Otherwise the pointer is released only when every step between succeeds (which is not a good assumption).

See https://github.com/JuliaIO/JpegTurbo.jl/blob/33d53e772cb3c11c0f49082d6493b06bec6cbaea/src/decode.jl#L105-L107 for an example